File: android/app/src/main/java/com/bluelink/android/domain/HistoryQuery.kt
package com.bluelink.android.domain

import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

enum class HistoryKind(val label: String) { ALL("全部"), TEXT("文本"), IMAGES("图片"), FILES("文件"), DATE("日期") }
enum class FileStatusFilter(val label: String) { ALL("全部状态"), ACTIVE("传输中"), COMPLETED("已完成"), INCOMPLETE("未完成"), FAILED("失败"), REJECTED("未接收"), CANCELED("已取消") }
enum class FileDirectionFilter(val label: String) { ALL("全部方向"), SENT("已发送"), RECEIVED("已接收") }

object HistoryQuery {
    val activeStatuses = setOf(TransferStatus.OFFERED, TransferStatus.QUEUED, TransferStatus.TRANSFERRING,
        TransferStatus.PAUSED, TransferStatus.REMOTE_PAUSED, TransferStatus.RESUMING, TransferStatus.VERIFYING, TransferStatus.COMMITTING)

    fun messages(items: List<ChatItem>, query: String, kind: HistoryKind, date: LocalDate? = null,
                 zone: ZoneId = ZoneId.systemDefault(), endDate: LocalDate? = date): List<ChatItem> = items.filter { item ->
        val needle = query.trim()
        val day = item.timestamp.atZone(zone).toLocalDate()
        (date == null || day >= date) && (endDate == null || day <= endDate) &&
        (needle.isEmpty() || item.text.contains(needle, ignoreCase = true) ||
            item.attachments.any { it.fileName.contains(needle, ignoreCase = true) }) && when (kind) {
            HistoryKind.ALL -> true
            HistoryKind.TEXT -> item.kind == ChatItemKind.TEXT
            HistoryKind.IMAGES -> item.attachments.any { it.isImage }
            HistoryKind.FILES -> item.attachments.any { !it.isImage }
            HistoryKind.DATE -> true
        }
    }.sortedWith(compareByDescending<ChatItem> { it.timestamp }.thenBy { it.id.toString() })

    fun files(items: List<TransferItem>, query: String, status: FileStatusFilter, direction: FileDirectionFilter,
              peerId: String?): List<TransferItem> = HistorySearch.files(items, FileQueryOptions(query, status, direction, peerId))

    fun showTimestamp(previous: Instant?, current: Instant, zone: ZoneId = ZoneId.systemDefault()): Boolean =
        previous == null || previous.atZone(zone).toLocalDate() != current.atZone(zone).toLocalDate() ||
            Duration.between(previous, current).toMinutes() >= 5
}

File: windows/BlueLink.App/Domain/HistoryQuery.cs
namespace BlueLink.Domain;

public enum HistoryKind { All, Text, Images, Files, Date }

public static class HistoryQuery
{
    // The first message of each day already has a date group header.
    public static bool HasMessageTimeGap(ChatItem current, ChatItem previous) =>
        current.CreatedAt.LocalDateTime.Date == previous.CreatedAt.LocalDateTime.Date &&
        current.CreatedAt - previous.CreatedAt >= TimeSpan.FromMinutes(10);

    public static bool Matches(ChatItem item, string query, HistoryKind kind, DateTime? date = null)
    {
        if (kind == HistoryKind.Text && (item.Kind != ChatItemKind.Text || !item.HasText)) return false;
        if (kind == HistoryKind.Images && item.Kind != ChatItemKind.Image && item.Attachments?.Any(a => a.IsImage) != true) return false;
        if (kind == HistoryKind.Files && item.Kind != ChatItemKind.File && item.Attachments?.Any(a => !a.IsImage) != true) return false;
        if (date is { } day && item.CreatedAt.LocalDateTime.Date != day.Date) return false;
        var term = query.Trim();
        return term.Length == 0 || item.Text.Contains(term, StringComparison.OrdinalIgnoreCase) ||
            item.Attachments?.Any(a => a.FileName.Contains(term, StringComparison.OrdinalIgnoreCase)) == true;
    }

    public static bool Matches(TransferItem item, string query, string status, string direction, string? peerId)
    {
        if (!string.IsNullOrEmpty(peerId) && !string.Equals(item.PeerId, peerId, StringComparison.OrdinalIgnoreCase)) return false;
        if (direction == "Outgoing" && !item.Outgoing || direction == "Incoming" && item.Outgoing) return false;
        if (status == "Active" && !item.IsActive || status == "Completed" && !item.IsCompleted || status == "Incomplete" && !item.IsRetryableTerminal || status == "Failed" && item.Status != TransferStatus.Failed ||
            status == "Rejected" && item.Status != TransferStatus.Rejected || status == "Canceled" && !item.IsCanceled) return false;
        return item.Name.Contains(query.Trim(), StringComparison.OrdinalIgnoreCase);
    }
}
